### ISSUE TITLE
Questions regarding vmagent configs

### ISSUE THREAD

**Issue by DaRealCapz** [VictoriaMetrics/VictoriaMetrics#5031]
### Is your question request related to a specific component?

VictoriaMetrics, vmagent

### Describe the question in detail

I want to limit the amount of data sent from scraped metrics between several vmagents and a VictoriaMetrics storage. After looking in the docs from vmagent i learned that metrics are pushed using the remote-write protocol.
Questions:
1- In the advanced usage [https://docs.victoriametrics.com/vmagent.html#advanced-usage](url) I noticed 2 similar flags, -pushmetrics.interval and -remoteWrite.flushInterval what is the difference between them and when are they best suited to be used?
2- Is there a way for me to change one of the previous named flags dynamically in runtime without having to rerun the system and changing the flag value?
3- When it comes to relabeling some metrics, if I have a vmagent scrapping metrics from a container is there a way to retrieve information from that container or virtual machine and relabel metrics with that information

### Troubleshooting docs

- [X] General - https://docs.victoriametrics.com/Troubleshooting.html
- [X] vmagent - https://docs.victoriametrics.com/vmagent.html#troubleshooting
- [ ] vmalert - https://docs.victoriametrics.com/vmalert.html#troubleshooting

**Comment by zekker6** [comment]
Hello @DaRealCapz 

> 1- In the advanced usage [https://docs.victoriametrics.com/vmagent.html#advanced-usage](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/url) I noticed 2 similar flags, -pushmetrics.interval and -remoteWrite.flushInterval what is the difference between them and when are they best suited to be used?

`-pushmetrics.interval` is used together with `-pushmetrics.url`:
```
  -pushmetrics.url array
     Optional URL to push metrics exposed at /metrics page. See https://docs.victoriametrics.com/#push-metrics . By default, metrics exposed at /metrics page aren't pushed to any remote storage
     Supports an array of values separated by comma or specified via multiple flags.
```
This flag is used to push vmagent metrics to the remote storage while `-remoteWrite.flushInterval` is used with `-remoteWrite.url`:
```
  -remoteWrite.url array
     Remote storage URL to write data to. It must support either VictoriaMetrics remote write protocol or Prometheus remote_write protocol. Example url: http://<victoriametrics-host>:8428/api/v1/write . Pass multiple -remoteWrite.url options in order to replicate the collected data to multiple remote storage systems. The data can be sharded among the configured remote storage systems if -remoteWrite.shardByURL flag is set. See also -remoteWrite.multitenantURL
     Supports an array of values separated by comma or specified via multiple flags.
```


>  2- Is there a way for me to change one of the previous named flags dynamically in runtime without having to rerun the system and changing the flag value?

No, currently there is no such option.


>  3- When it comes to relabeling some metrics, if I have a vmagent scrapping metrics from a container is there a way to retrieve information from that container or virtual machine and relabel metrics with that information

In order to do this you need to use [service discovery](https://docs.victoriametrics.com/sd_configs.html) of containers. Service discover add meta information about the container and this meta infromation can be used during relabeling later on.

**Comment by DaRealCapz** [comment]
First of all I want to thank you for the reply, but on the first question it was unclear for me when should I use what flag. For instance in the following system what flag should I use to optimize the metric pushing frequency?

<img width="402" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/71236784/73410563-0dc2-4266-91a1-28164baa27ce">

regarding question 2, is there a way for it to be developed by me? Or does the underlying code not allow for it?

**Comment by zekker6** [comment]
> First of all I want to thank you for the reply, but on the first question it was unclear for me when should I use what flag. For instance in the following system what flag should I use to optimize the metric pushing frequency?

For this case you need to use `-remoteWrite.flushInterval`.


> regarding question 2, is there a way for it to be developed by me? Or does the underlying code not allow for it?

This will require significant changes to the code base. Currently, packages define flags which are used inside those packages. In order to be able to change values for those flags it will be required to move all flags into some kind of configuration, add initialization and re-reading this config to apply modified options.

**Comment by DaRealCapz** [comment]
Ok, so if I understood it, the only way to change `-remoteWrite.flushInterval` is to fully reset the runtime? Any tips as to what it would take, code wise, to allow for it to be dynamically changed?

**Comment by zekker6** [comment]
Yes, you need to restart vmagent to apply new values. Based on my expirience, I would say that most likely after initial configuration and adjustment of value for this flag it will never be modified in the future.

Code wise, you will need to update vmagent to be able to re-initialize internal queues/flushing mechanism once this value changes and to move this flag to configuration file or environment variable.

**Comment by air3ijai** [comment]
```
  -pushmetrics.interval duration
     Interval for pushing metrics to every -pushmetrics.url (default 10s)

  -remoteWrite.flushInterval duration
     Interval for flushing the data to remote storage. This option takes effect only when less than 10K data points per second are pushed to -remoteWrite.url (default 1s)
```

How we can use `flushInterval` to save the bandwidth?

Even if we set it to `300s`, data is coming to Grafana almost instantly.

**Comment by zekker6** [comment]
@air3ijai 
> How we can use `flushInterval` to save the bandwidth?

`flushInterval` will not help to reduce the bandwidth. 
Data is getting into queue after being scraped or ingested via push protocols. `flushInterval` defines a time interval between attempts to push the data if data haven't been flushed yet.
There are two more options which control this behaviour:
- `remoteWrite.maxBlockSize` - the block size which is pushed at once
- `remoteWrite.maxRowsPerBlock` - number of rows which are pushed at once

If currently buffered data size will exceed any of these options it will be flushed before `flushInterval`.


If you need to save bandwidth a better way would be to ensure that you're using vmagent and the destination with version which supports [VictoriaMetrics remote-write protocol](https://docs.victoriametrics.com/vmagent/#victoriametrics-remote-write-protocol) (it was added in `v1.90.0`).

**Comment by air3ijai** [comment]
We are running VictoriaMetrics in in Kubernetes and VMAgent on VMs
```
victoriametrics/victoria-metrics:v1.99.0
victoriametrics/vmagent:v1.99.0
```

Adjusted configuration as following and will check how it performs
```
remoteWrite.forceVMProto         = true
remoteWrite.flushInterval        = 300
remoteWrite.maxBlockSize         = 15MB
remoteWrite.maxRowsPerBlock      = 30000
remoteWrite.vmProtoCompressLevel = 10
```

**Comment by jiekun** [comment]
Close due to inactivity.
